packager/signer/validation.go

```diff
@@ -0,0 +1,165 @@
+// Auxiliary functions for use by Signer.
+package signer
+
+import (
+	"mime"
+	"net/http"
+	"net/url"
+	"regexp"
+
+	"github.com/ampproject/amppackager/packager/util"
+	"github.com/pkg/errors"
+	"github.com/pquerna/cachecontrol"
+)
+
+func parseURL(rawURL string, name string) (*url.URL, *util.HTTPError) {
+	if rawURL == "" {
+		return nil, util.NewHTTPError(http.StatusBadRequest, name, " URL is unspecified")
+	}
+	ret, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, util.NewHTTPError(http.StatusBadRequest, "Error parsing ", name, " URL: ", err)
+	}
+	if !ret.IsAbs() {
+		return nil, util.NewHTTPError(http.StatusBadRequest, name, " URL is relative")
+	}
+	// Evaluate "/..", by resolving the URL as a reference from itself.
+	// This prevents malformed URLs from eluding the PathRE protections.
+	ret = ret.ResolveReference(ret)
+	return ret, nil
+}
+
+func regexpFullMatch(pattern string, test string) bool {
+	// This is how regexp/exec_test.go turns a partial pattern into a full pattern.
+	fullRe := `\A(?:` + pattern + `)\z`
+	matches, _ := regexp.MatchString(fullRe, test)
+	return matches
+}
+
+func urlMatches(url *url.URL, pattern util.URLPattern) error {
+	if url.Opaque != "" {
+		return errors.New("URL is opaque")
+	}
+	if url.User != nil {
+		return errors.New("URL contains user")
+	}
+	if !regexpFullMatch(*pattern.PathRE, url.EscapedPath()) {
+		return errors.New("PathRE doesn't match")
+	}
+	for _, re := range pattern.PathExcludeRE {
+		if regexpFullMatch(re, url.EscapedPath()) {
+			return errors.Errorf("PathExcludeRE matches: %s", re)
+		}
+	}
+	if !regexpFullMatch(*pattern.QueryRE, url.RawQuery) {
+		return errors.New("QueryRE doesn't match")
+	}
+	return nil
+}
+
+func schemeMatches(actual string, expecteds []string) bool {
```

Comment:
Instead of `expecteds` perhaps `schemes` or `expected_schemes`, though I know Go prefers brevity with naming.

Reply:
Done.